samples/debug/memfault/src/main.c

```diff
@@ -6,22 +6,37 @@
 
 #include <zephyr/kernel.h>
 #include <stdio.h>
+#include <zephyr/sys/reboot.h>
 #include <zephyr/net/socket.h>
-#include <dk_buttons_and_leds.h>
-
+#include <zephyr/net/conn_mgr_connectivity.h>
+#include <zephyr/net/conn_mgr.h>
 #include <memfault/metrics/metrics.h>
 #include <memfault/ports/zephyr/http.h>
 #include <memfault/core/data_packetizer.h>
 #include <memfault/core/trace_event.h>
+#include <dk_buttons_and_leds.h>
 
 #include <zephyr/logging/log.h>
-
-#include "network_common.h"
+#include <zephyr/logging/log_ctrl.h>
 
 LOG_MODULE_REGISTER(memfault_sample, CONFIG_MEMFAULT_SAMPLE_LOG_LEVEL);
 
+/* Macros used to subscribe to specific Zephyr NET management events. */
+#define L4_EVENT_MASK	      (NET_EVENT_L4_CONNECTED | NET_EVENT_L4_DISCONNECTED)
+#define CONN_LAYER_EVENT_MASK (NET_EVENT_CONN_IF_FATAL_ERROR)
+
 static K_SEM_DEFINE(nw_connected_sem, 0, 1);
 
+/* Macro called upon a fatal error, reboots the device. */
+#define FATAL_ERROR()                                                                              \
+	LOG_ERR("Fatal error! Rebooting the device.");                                             \
+	LOG_PANIC();                                                                               \
+	IF_ENABLED(CONFIG_REBOOT, (sys_reboot(0)))
```

Comment:
I suggest that we don't reboot directly in this sample, as the fatal errors will not cause coredumps and therefore not be reported to Memfault. I think we should use `__ASSERT(...)` insted.